### ISSUE TITLE
Incorrect type inference for TypedDict declared variable in loops

### ISSUE THREAD

**Issue by Voldemat** [astral-sh/ty#4206]
### Summary

example that works without errors:
```python
from typing import TypedDict, Required

class Arguments(TypedDict):
    value: Required[int]

values: Arguments
values = {
    "value": 1,
}
```

Example that throws error:
```python
from typing import TypedDict, Required

class Arguments(TypedDict):
    value: Required[int]

while True:
    values: Arguments
    values = {
        "value": 1,
    }
```
Logs:
```
error[[invalid-declaration](https://ty.dev/rules#invalid-declaration)]: Cannot declare type `Arguments` for inferred type `dict[str, int]`
 --> check.py:7:5
  |
7 |     values: Arguments
  |     ^^^^^^
```

### Version

ty 0.0.69 (1dca8443e 2026-08-06)

**Comment by carljm** [comment]
Thanks! Definitely a bug; we need to apply the declaration context of `Arguments` to inference of the dict assignment before trying to validate that the declaration is valid for the inferred type.

We should fix this but it might not be top priority, since moving the declaration out of the loop should generally be a valid and easy fix.

**Comment by michtesar** [comment]
hey @carljm can I take this one?

**Comment by carljm** [comment]
Sure; I'm not sure how involved the fix will be, but you're certainly welcome to work on it.

**Comment by carljm** [comment]
Sorry @michtesar, looks like @charliermarsh got there first :)
